trusted-firmware-m: Fix sign images to allow FOTA

nandojve opened the pull request:
The `trusted-firmware-m` requires specific instructions when signing the images to allow local development and FOTA. This PR extends the sign process inside Zephyr to generate the hex files with the confirmed image to load firmware for development and production purposes. It add the missing binary sign to generate the correct non confirmed image to allow FOTA upgrade. These images now respect the TF-M/FWU state machine FOTA is now possible.

These was tested on STM32U5A5xx SoC with custom OTP and Keys.

Fixes #94204

nandojve committed:
trusted-firmware-m: Set --align when signing

The current version of TF-M script that sign MCUboot image uses a
default alignment of 1. This value varies between flash devices
and not all accept the default 1. This improve the script picking
the write-block-size property from the current flash controller
and pass as the --align parameter when signing an image.

Note: This solution works out-of-box for the vast majority of
devices in the Zephyr tree and an exception will throw when
a device is not supported.

Signed-off-by: BUDKE Gerson Fernando <gerson.budke@leica-geosystems.com>

nandojve committed:
trusted-firmware-m: Set --max-sectors when signing

The --max-sectors option helps catch problems with flash overlap when
merging images. If there is a misalignment in flash partitions, the
merge process usually fails. This uses information from Zephyr flash
partitions and the flash controller to automatically determine the
max sectors value and apply it when signing an image.

Signed-off-by: BUDKE Gerson Fernando <gerson.budke@leica-geosystems.com>

nandojve committed:
trusted-firmware-m: Define header and trailer options

The current behavior when signing an image is to always set --pad and
--pad-header for all images unless TFM_USE_NS_APP is set. This does not
allow for easy creation of signed images for FOTA applications. Rewrite
the PAD parameter as HEADER and TRAILER to simplify the setup of more
signing options.

Another important reason for this change is that the NS image, when
signed without --pad, runs on the hardware but does not perform the
MCUboot test, and the FWU never upgrades the image. This fixes the NS
image signing process to correctly support TF-M FWU using the PSA API
functions.

Signed-off-by: BUDKE Gerson Fernando <gerson.budke@leica-geosystems.com>

nandojve committed:
trusted-firmware-m: Set --confirm when signing

The current behavior when signing an image adds --pad but does not
confirm the image. This appears to be a mistake, as the user should
inspect the image status in the Firmware Upgrade software. If an image
is not --confirmed, the FSM cannot infer the correct states. This sets
the image as confirmed to resolve the issue.

Signed-off-by: BUDKE Gerson Fernando <gerson.budke@leica-geosystems.com>

nandojve committed:
trusted-firmware-m: Make hex files variables explicit

Make variables that define output files explicitly include 'HEX' in the
name. This refactoring step allows for the introduction of BIN file
generation.

Signed-off-by: BUDKE Gerson Fernando <gerson.budke@leica-geosystems.com>

nandojve committed:
trusted-firmware-m: Create multi image bin files

A fundamental use of Trusted Firmware-M is to provide security for
IoT applications, where firmware upgrades (FOTA) are almost always
mandatory. The current file signing process does not produce the
necessary binaries for multi-image S/NS FWU, since hex images are
not suitable for this use case. This introduces the missing signed
binary files for use by the FWU partition. The changes were tested
in multi-image FWU scenarios, and support for single-image scenarios
can be easily added in the future.

Signed-off-by: BUDKE Gerson Fernando <gerson.budke@leica-geosystems.com>

lindblandro commented on modules/trusted-firmware-m/CMakeLists.txt:
```diff
@@ -533,7 +537,7 @@ if (CONFIG_BUILD_WITH_TFM)
 
   elseif(CONFIG_TFM_MCUBOOT_IMAGE_NUMBER STREQUAL "1")
-    tfm_sign(sign_cmd S_NS TRUE ${S_MAX_SECTORS} ${S_NS_FILE} ${S_NS_SIGNED_FILE})
+    tfm_sign(sign_cmd S_NS TRUE TRUE ${S_MAX_SECTORS} ${S_NS_FILE} ${S_NS_SIGNED_FILE})
```
This is becoming increasingly convoluted to read. How about changing it to use `cmake_parse_arguments()`? Then calls would look something like:
```cmake
tfm_sign(OUTPUT imgtool_cmd 
  SUFFIX "S_NS"
  ALIGN "${flash_align}"
  MAX_SECTORS "${S_MAX_SECTORS}"
  INPUT_FILE "${S_NS_FILE}"
  OUTPUT_FILE "${S_NS_SIGNED_FILE}"
  PAD_HEADER 
  PAD_TRAILER)
``` 
I think this would be more idiomatic CMake, make the code more readable and make it easier to add new options later (if necessary).

nandojve replied:
Done in e6cfcf7

arnebdr commented:
I have tested the patch, works for my project using U585. I think its a good solution

nandojve committed:
boards: arm: mps2: Fix NS flash layout

The mps2/an521/cpu0/ns define flash layouts in tf-m to allow
CONFIG_TFM_MCUBOOT_IMAGE_NUMBER be 1 or 2. In the Zephyr project when
building the samples the value selected is 2.

The layout changes are necessary to allow use the --max-sectors options
when signing the images. It ensures that flash layout is respected. To
allow this the compatible "soc-nv-flash" was added in the reserved
memory and the fixed-partitions were defined.

Signed-off-by: BUDKE Gerson Fernando <gerson.budke@leica-geosystems.com>

nandojve committed:
boards: st: stm32l562e_dk: Move external partitions

The external partitions are defined dependent from S and NS images. This
move the external partitions from common to the S image. The NS image
will be defined in future to allow correct usage of MCUboot.

Signed-off-by: BUDKE Gerson Fernando <gerson.budke@leica-geosystems.com>

nandojve committed:
boards: st: stm32l562e_dk: ns: Align partitions

The partitions for the stm32l562e_dk/stm32l562xx/ns board are not aligned
with TF-M. This fixes the partition alignment.

Signed-off-by: BUDKE Gerson Fernando <gerson.budke@leica-geosystems.com>

lindblandro commented on modules/trusted-firmware-m/CMakeLists.txt:
I can't quite follow your reasoning here. Without `--confirm` the application must perform `psa_fwu_accept()`, which seems like a reasonable requirement. This now unconditionally marks all images as confirmed making `psa_fwu_accept()` a no-op. The safest default would be to require the application to explicitly confirm / accept the new image. 

Or am I missing something here? :) Did you mean that `psa_fwu_query()` doesn't report the correct image states if the image is not confirmed?

nandojve replied:
There are a few points:

1- The HEX file is used only in production and development phase. It means that it is always flashed in the board using a JTAG/SWD tool. So, we can assume that in development phase the image is always good and for production I can not imagine that someone would like to flash an image that it is not good.
2- When flashing direct in the slot there will be no mechanism to mark the image as good.
3- When you do not use FWU there will be no routine to mark the image as good.

What I understood is, to the FWU works the image should be marked correctly to allow swap, when appropriated. The binary image MUST not be confirmed because system should always mark the image as good / bad. The system can only mark the image as good if FWU is present and it marks both S and NS at same time.

nordicjm commented on boards/st/stm32l562e_dk/stm32l562e_dk_stm32l562xx_ns.dts:
```diff
@@ -41,24 +41,64 @@
 		};
 
-		/* Secure image primary slot */
+		scratch_partition: partition@11000 {
+			reg = <0x00011000 DT_SIZE_K(8)>;
+		};
```
not a problem with the PR but just to point out that assuming the 70% of the application space here is used, that would be about ~330KiB and one single firmware update would do 41 or 82 (can't remember if it uses it when writing in one or both directions) flash erase/write cycles on this section here, meaning the flash endurance cycles will go down far faster than the rest of the device (for a 10k endurance flash device that means ~123 or ~246 firmware updates)

nandojve replied:
I complete agreed with you here and I'll propose a patch to mainline.

nandojve commented:
* patch set tries to address the CI failings
* rebase

wearyzen reviewed (dismissed):
I haven't looked at the PR but just adding a -1 so that the MPS2/MPS3 changes don't get merged. I'll have a look at this next week once I am done with my current high priority task.

BTW, @nandojve, currently mps4/**/fvp/ns boards are not enabled because TF-M doesn't generate few required hex files, would this PR help with that? if you would like to try you can add below 1 line [here](https://github.com/zephyrproject-rtos/zephyr/blob/1cb0b6a7ffa5f76daa873ec4345e8fe35380c5b3/modules/trusted-firmware-m/Kconfig.tfm#L19) 
```
default "arm/mps4/corstone320/fvp"   if BOARD_MPS4_CORSTONE320_FVP_NS
```

nandojve commented:
Hi @wearyzen ,

Are you maintaining the ARM platform ?

> I haven't looked at the PR but just adding a -1 so that the MPS2/MPS3 changes don't get merged. I'll have a look at this next week once I am done with my current high priority task.

I have a product that require TF-M + FWU ASAP in mainline. This changes was already proved that makes FWU work by community inclusive. I'm only trying to fix mps2/3 because are in CI but I'm fine to exclude then from tests because those bring 0 value at moment, if they build.

> BTW, @nandojve, currently mps4/**/fvp/ns boards are not enabled because TF-M doesn't generate few required hex files, would this PR help with that? if you would like to try you can add below 1 line [here](https://github.com/zephyrproject-rtos/zephyr/blob/1cb0b6a7ffa5f76daa873ec4345e8fe35380c5b3/modules/trusted-firmware-m/Kconfig.tfm#L19)
> 
> ```
> default "arm/mps4/corstone320/fvp"   if BOARD_MPS4_CORSTONE320_FVP_NS
> ```

I'm not sure because flash layout from ARM does not provide correct info at all. The layout comment says single image but CONFIG_TFM_MCUBOOT_IMAGE_NUMBER=2.

https://discord.com/channels/1106321706588577904/1106321982900945017/1405922987491983581

nandojve commented:
> > Hi @wearyzen ,
> > Are you maintaining the ARM platform ?
> 
> Yes I am.
> 
> > >

Great!

> That would be breaking the mps2/mps3 ns boards isn't it?

It should not break, see here https://github.com/zephyrproject-rtos/zephyr/pull/94470#issuecomment-3190965494 running locally with the changes. So, the mps2 is working normally and it is passing on the tests. The mps3 is not working in mainline which is creating problem in this PR at moment. The mps4 I did not touched.

> if you could you change the code to have the current behavior for mps2/mps3/mps4 and the new change only for others then I don't have any issue but with current PR, I will still need some time to understand the changes to approve it.

I don't mind to update those. In fact, it will be nice to have those working because then we can bring again the FWU sample in mainline with tests in future.

@wearyzen, if you can bring the mpsX to live in mainline we can work together to make sure everything is Ok.

nandojve committed:
modules: tf-m: Kconfig.tfm: Update TFM_BOARD

Reorder TFM_BOARD entries by vendor name for improved clarity.

Signed-off-by: BUDKE Gerson Fernando <gerson.budke@leica-geosystems.com>

wearyzen reviewed (commented):
added a few comments but the mps2/mps3 changes LGTM

nandojve committed:
boards: arm: mps3: Remove mps3/corstone300/an547/ns QEMU

Exclude mps3/corstone300/an547/ns from PSA tests because QEMU does not
model the QSPI flash in MPS3 boards as real QSPI flash, but only as
simple ROM. Attempting to rewrite the flash from the guest will fail.

https://github.com/zephyrproject-rtos/zephyr/pull/94470#issuecomment-3197729501

Signed-off-by: BUDKE Gerson Fernando <gerson.budke@leica-geosystems.com>

nandojve committed:
docs: migration-guide-4.3: Add TF-M sign note

Add a note about BL2 (MCUboot) signing updates when the board is built
as TF-M NS.

Signed-off-by: BUDKE Gerson Fernando <gerson.budke@leica-geosystems.com>

nandojve commented on samples/psa/its/sample.yaml:
```diff
@@ -16,6 +16,8 @@ tests:
     extra_args: EXTRA_CONF_FILE=overlay-tfm.conf
     tags:
       - trusted-firmware-m
+    platform_exclude:
```
Your suggestion worked : )

nandojve commented on modules/trusted-firmware-m/CMakeLists.txt:
```diff
@@ -471,7 +476,15 @@ if (CONFIG_BUILD_WITH_TFM)
+  # Calculate the maximum number of sectors necessary to store the image.
+  dt_nodelabel(s_partition_node NODELABEL "slot0_partition")
+  dt_nodelabel(ns_partition_node NODELABEL "slot0_ns_partition")
+  dt_reg_size(s_partition_size PATH ${s_partition_node})
+  dt_reg_size(ns_partition_size PATH ${ns_partition_node})
+  math(EXPR S_MAX_SECTORS "${s_partition_size} / ${flash_erase_block_size}")
+  math(EXPR NS_MAX_SECTORS "${ns_partition_size} / ${flash_erase_block_size}")
```
I don't like this suggestion. I understand the reasoning but in this particular case, the boards that do not follow the naming are wrong and need to be fixed.

The problem that I want to avoid is that blocks values are different depending on the configuration:

```
#if (MCUBOOT_IMAGE_NUMBER == 1)
/* The maximum number of status entries supported by the bootloader. */
#define MCUBOOT_STATUS_MAX_ENTRIES ((FLASH_S_PARTITION_SIZE + \
                                     FLASH_NS_PARTITION_SIZE) / \
                                    FLASH_AREA_SCRATCH_SIZE)
/* Maximum number of image sectors supported by the bootloader. */
#define MCUBOOT_MAX_IMG_SECTORS    ((FLASH_S_PARTITION_SIZE + \
                                     FLASH_NS_PARTITION_SIZE) / \
                                    FLASH_AREA_IMAGE_SECTOR_SIZE)

#elif (MCUBOOT_IMAGE_NUMBER == 2)

#if (FLASH_S_PARTITION_SIZE > FLASH_NS_PARTITION_SIZE)
#define FLASH_MAX_PARTITION_SIZE FLASH_S_PARTITION_SIZE
#else
#define FLASH_MAX_PARTITION_SIZE FLASH_NS_PARTITION_SIZE
#endif

/* The maximum number of status entries supported by the bootloader. */
#define MCUBOOT_STATUS_MAX_ENTRIES (FLASH_MAX_PARTITION_SIZE / \
                                    FLASH_AREA_SCRATCH_SIZE)

/* Maximum number of image sectors supported by the bootloader. */
#define MCUBOOT_MAX_IMG_SECTORS           ((FLASH_MAX_PARTITION_SIZE) / \
                                           FLASH_AREA_IMAGE_SECTOR_SIZE)

#else /* MCUBOOT_IMAGE_NUMBER > 2 */
#error "Only MCUBOOT_IMAGE_NUMBER 1 and 2 are supported!"
#endif /* MCUBOOT_IMAGE_NUMBER */
```

Guessing these block my not align with the reality, see changes in MPS2. Now with these changes, the values are compared with the flash_layout
I imagine that it is clear this should be defined correctly to avoid problems. Reviewing my patch I can still see that some issues can be present and I'll fix those.

wearyzen replied:
Let me try to explain again...
Today, with or without CONFIG_TFM_BL2 enabled, there are board that do not have any of these nodes and manage to work just fine. This change breaks these boards and so cannot be accepted as is. I suggested to have the defaults because they won't break these boards and they are the same default that the imgtool considers today for signing.

nordicjm replied:
If you don't have bl2 then you don't have signing or signed images or imgtool

wearyzen replied:
yeah you are right, I missed that.. updated my comment.

nandojve committed:
trusted-firmware-m: Use cmake_parse_arguments in tfm_sign

Use cmake_parse_arguments() for more idiomatic code. This makes the
code more readable and easier to extend with new options.

Signed-off-by: BUDKE Gerson Fernando <gerson.budke@leica-geosystems.com>

nandojve commented:
* rebase
* address small issues
* add cmake parser function

wearyzen commented on modules/trusted-firmware-m/CMakeLists.txt:
```diff
@@ -494,65 +546,72 @@ if (CONFIG_BUILD_WITH_TFM)
     else()
-      tfm_sign(sign_cmd_ns NS FALSE ${NS_APP_FILE} ${NS_SIGNED_FILE})
+      tfm_sign(sign_cmd_ns_hex NS FALSE TRUE TRUE ${NS_MAX_SECTORS} ${NS_APP_FILE} ${NS_SIGNED_HEX_FILE})
+      tfm_sign(sign_cmd_ns_bin NS FALSE TRUE FALSE ${NS_MAX_SECTORS} ${NS_APP_FILE} ${NS_SIGNED_BIN_FILE})
```
as per https://github.com/zephyrproject-rtos/zephyr/pull/94470/files#diff-dc28570be9d838c291b4be7f1ea2436d81b17f095892a52869cd3cd88317c599R560-R566 NS_APP_FILE is a hex file, I haven't tested this as well but IIUC the format of input and output files should be same (both hex or bot bin) and that's why I said this doesn't look right.

nandojve replied:
I'm checking this internally.

nandojve replied:
I asked to a team evaluate this. I think next week we have a clear picture.

nandojve replied:
This was fixed.

nandojve commented:
* fix typo

etienne-lms reviewed (commented):
I'm fine with the 2 STM32 specific commits: "boards: st: stm32l562e_dk: ...".
I find the overall idea of this P-R consistent and legitimate (FWIW).

nordicjm commented on modules/trusted-firmware-m/CMakeLists.txt:
```diff
@@ -444,6 +444,33 @@ if (CONFIG_BUILD_WITH_TFM)
 
+  dt_chosen(chosen_flash PROPERTY "zephyr,flash")
+
+  dt_prop(flash_write_block_size PATH ${chosen_flash} PROPERTY write-block-size)
```
For other dt functions there is the REQUIRED flag and if you fail earlier you get a better message, e.g. https://github.com/mcu-tools/mcuboot/pull/2427 but there isn't one for dt_chosen so it just means the dt_chosen() part will always succeed but the dt_prop one will throw an error that you haven't provided a path instead of saying that there isn't a chosen node

tomi-font commented on doc/releases/migration-guide-4.3.rst:
```diff
@@ -172,5 +172,16 @@ Silabs
+  tf-m NS must have their flash layout with the flash controller information.
+  This ensure that when signing the hex/bin files all the details will be
+  present in the S and NS image. This change fixes the image details to allow
+  the FWU state machine be correct and allow FOTA.
+
+  For more details: https://github.com/zephyrproject-rtos/zephyr/pull/94470
```
```suggestion
  the FWU state machine to be correct and allow FOTA. (:github:`94470`)
```

tomi-font commented on doc/releases/migration-guide-4.3.rst:
```diff
@@ -172,5 +172,16 @@ Silabs
+
+* The signing process for BL2 (MCUboot) was updated. The boards that run using
+  tf-m NS must have their flash layout with the flash controller information.
+  This ensure that when signing the hex/bin files all the details will be
```
```suggestion
  This ensures that when signing the hex/bin files all the details will be
```

nandojve commented on modules/trusted-firmware-m/CMakeLists.txt:
```diff
@@ -579,12 +583,12 @@ if (CONFIG_BUILD_WITH_TFM)
+      tfm_sign(sign_cmd_ns NS TRUE TRUE ${S_NS_MAX_SECTORS} ${NS_APP_FILE} ${NS_SIGNED_FILE})
     else()
-      tfm_sign(sign_cmd_ns NS FALSE ${S_NS_MAX_SECTORS} ${NS_APP_FILE} ${NS_SIGNED_FILE})
+      tfm_sign(sign_cmd_ns NS FALSE TRUE ${S_NS_MAX_SECTORS} ${NS_APP_FILE} ${NS_SIGNED_FILE})
```
You mean `CONFIG_TFM_USE_NS_APP=n`, this was fixed.

nandojve commented:
* rebase
* add check to `CONFIG_TFM_BL2` to avoid issues when BL2 is disabled.
* minor issues addressed

nandojve commented:
Hi,

The CI is happy now and I think I address almost all comments.
I invite you to revisit the PR to check.

nandojve committed:
boards: arm: mps3: Fix NS flash layout and SRAM size

The mps3/foo/ns configuration defines flash layouts in TF-M to support
CONFIG_TFM_MCUBOOT_IMAGE_NUMBER values of 1 or 2. In the Zephyr
project, when building samples, the selected value is 2.

The layout changes are necessary to support the --max-sectors option
when signing images, ensuring that the flash layout is respected. To
enable this, the compatible "soc-nv-flash" was added to the reserved
memory, and fixed-partitions were defined.

Additionally, the ISRAM was redefined to expose the correct size and
values for both S and NS firmware, clarifying memory selection for the
user.

For example, see general details in:
https://git.trustedfirmware.org/plugins/gitiles/TF-M/trusted-firmware-m.git/%2B/refs/heads/main/platform/ext/target/arm/mps3/corstone310/common/partition/region_defs.h#116
https://git.trustedfirmware.org/plugins/gitiles/TF-M/trusted-firmware-m.git/%2B/refs/heads/main/platform/ext/target/arm/mps3/corstone310/common/config.cmake#13

Note:
 - Not all mps3 ISRAM have the same size and design should take that
   in consideration.

Signed-off-by: BUDKE Gerson Fernando <gerson.budke@leica-geosystems.com>

nandojve commented:
* fix https://github.com/zephyrproject-rtos/zephyr/pull/94470/files#r2284854508

nandojve commented on modules/trusted-firmware-m/CMakeLists.txt:
```diff
@@ -475,9 +477,26 @@ if (CONFIG_BUILD_WITH_TFM)
     endif()
+
+    # Calculate the maximum number of sectors necessary to store the image.
+    dt_nodelabel(s_partition_node NODELABEL "slot0_partition" REQUIRED)
```
Hi @wearyzen ,

The `dt_nodelabel(s_partition_node NODELABEL "slot0_partition" REQUIRED)` is now inside a block that evaluate only when `if (CONFIG_TFM_BL2)` line 447 is valid.

This means that if user request a BL2, it will requires the `slot0_foo_partition` devicetree description. In Zephyr I think the convention is to use the names that I defined here in the scripts to collect the necessary content. The partitions were declared as REQUIRED, as suggestions from @nordicjm, to give a good error message to users.

I added a description in the migration guide that will inform users that they need to have the flash layout, including the flash controller information. I added in the migration text the PR number so people can easy check the history and see the changes made here.

Look, I don't feel comfortable to keep loop holes in mainline that lead to wrong configurations. For instance, imagine that ARM has those wrong configs in house (I'm saying that it is wrong OK - it is just an example). Then, on their CI, everything seems to be OK, but in fact, it could easy lead to bugs like this https://review.trustedfirmware.org/c/TF-M/trusted-firmware-m/+/42130 . Just because there is no description which create an image that user can test up/down upgrades.

For me the ultimate question is: why are we supposed to let people and companies to use wrong configs that induce to problems ? Instead, shouldn't the fix somehow pop up on their knowledge ?

wearyzen replied:
@nordicjm I think all your questions are related to firmware update which MPS3/MPS4 board do not support atm.
When I say they are working today, I mean you can build these boards and run on FVP simulation platform without any issue and test all the features of Cortex-m55/m85 with Ethos-U.
My concern is that if this PR gets merged it will break the build for MPS4 boards.

wearyzen replied:
I think the confusion here comes from assuming that MPS4 uses MCUboot. It doesn’t, the BL2 in use today comes directly from TF-M, and its flash layout is defined inside the TF-M repo (flash_layout.h, etc.), not from Zephyr’s devicetree.

Also worth noting, FWU isn’t the only purpose of BL2. It’s responsible for secure boot as well, which is something we can test on this board. That’s why BL2 is enabled, even without MCUboot or firmware update support.

Hope that clears things up.

nordicjm replied:
boards with 0 testing should not be added to zephyr, it's a requirement that hello world and philosophers work and using the board targets as they are in main and probably from the moment they were added right until current day, they do not and have never worked

nordicjm replied:
I refer you to https://docs.zephyrproject.org/latest/contribute/bin_blobs.html#limited-scope for which it states:
```
The project maintainers define an open source test suite that an upstream target must be able to pass using only open source software included in the mainline distribution and its modules. The open source test suite currently consists of:

    samples/philosophers

    tests/kernel

The scope of this test suite may grow over time. The goal is to specify tests for a minimal feature set which must be supported via open source software for any target with upstream Zephyr support.

At the discretion of the release team, the project may remove support for a hardware target if it cannot pass this test suite.
```
the mps4 board targets do not meet this minimum requirement, and could be scheduled for removal

nandojve replied:
Hi @wearyzen , @nordicjm ,

I think we should now focus in the real issue that let us to this discussion. The issue happens because a board (or many boards) do not follow the correctly memory definition.

In this case (I can be wrong in the terms, but I'l try to explain):

1- A Zephyr application can define the full flat memory using the address from "flash0" definition inside the flash controller or;
2- A Zephyr application must define a flash layout - partitions

In [1] it must not be allowed to use tricks to allow the app to run. Offsets should not to be allowed because infrastructure assumes that code will be compiled to start from that exact point were the reset will try to fetch the content "address 0x0". If user scenario require a different address that than, he should use [2].

In [2], user is free to define what they need. This is inclusive bootloader independent.

Based on the above, a Zephyr + TF-M application has only 2 options to define the memory region.

The option [1] will produce a merged hex file, no signing etc and the option [2] require partitions. The zephyr infrastructure will do the job to accommodate the content in the right place and call the necessary functions to generate the final image.

At moment, a Zephyr + TF-M application + bootloader mandate flash layout. This will imply that when a bootloader is involved (no matter what bootloader is) a flash layout is necessary anyway. I think this is the first **mistake** in many ARM boards at moment, some of those I'm proposing a fix in this PR inclusive. The boards that do not comply should be excluded from CI - minimum - and that should not be a justification to block a PR.

The second topic, having an independent bootloader (not MCUboot) in the TF-M CMake scripts is complete fair. However, this bootloader SHOULD be supported in Zephyr, at least, before a blocking claim be made.

That said, I'm closing this conversation because my goal is "force" compliance. The compliance will allow us to improve our tests and make sure the products that are made with Zephyr are reliable and secure. I do not oppose to have support to any other bootloader to be used with TF-M or with Zephyr itself but the support should be made in the mainstream to allow us to test those scenarios too. I'm not in favor to over engineer due to downstream scenarios because they are, usually, counter productive.

JordanYates commented on modules/trusted-firmware-m/CMakeLists.txt:
```diff
@@ -493,14 +493,18 @@ if (CONFIG_BUILD_WITH_TFM)
   endif()
 
-  function(tfm_sign OUT_ARG SUFFIX HEADER TRAILER MAX_SECTORS INPUT_FILE OUTPUT_FILE)
+  function(tfm_sign OUT_ARG SUFFIX HEADER TRAILER CONFIRM MAX_SECTORS INPUT_FILE OUTPUT_FILE)
```
Setting confirmed by default is incorrect for merged images, where the NS application can modify the image trailer for the combined image. That the secure application doesn't do any confirmation sounds like a problem to solve in TF-M, not by marking all OTA images as confirmed by default

nandojve replied:
Hi @JordanYates , 
There are small things in the PR to be complete. Could you check if my explanations are reasonable ?

JordanYates reviewed (changes requested):
Change request is for the `--confirm` commit.

General comments:
This PR seems to contain a whole bunch of extra changes not related to the core "Fix sign images for FOTA".
The actual fixes can be reduced to approximately 10 lines of changes, which I have been using successfully for over a year:
https://github.com/Embeint/zephyr/commit/6b98317f1573f7a3585ded6b89ff99e2291ea8a5
https://github.com/Embeint/zephyr/commit/eecaecf62faa7e673842c095bd219ef7349d7abe
https://github.com/Embeint/zephyr/commit/9051463f699e22b43132163a49cb6809d70c4723

nandojve commented:
> Change request is for the `--confirm` commit.
> 
> General comments: This PR seems to contain a whole bunch of extra changes not related to the core "Fix sign images for FOTA". The actual fixes can be reduced to approximately 10 lines of changes, which I have been using successfully for over a year: [Embeint@6b98317](https://github.com/Embeint/zephyr/commit/6b98317f1573f7a3585ded6b89ff99e2291ea8a5) [Embeint@eecaecf](https://github.com/Embeint/zephyr/commit/eecaecf62faa7e673842c095bd219ef7349d7abe) [Embeint@9051463](https://github.com/Embeint/zephyr/commit/9051463f699e22b43132163a49cb6809d70c4723)

Hi @JordanYates ,

Thank you for the inputs. I'll ask you to document here your scenario in details to get the whole picture. It will be nice to compare your setup with ours to make sure we are talking about same things. Then we can proper fix as many use cases possible.

nandojve commented:
* rebase due to #95295 fix.

nordicjm reviewed (commented):
some nits. needs @tomi-font to check, maybe @d3zd3z

nandojve commented:
* Fix nits
* Rollback the TRAILER in the NS HEX file. I can confirm that when no --pad is in the hex file the image will boot as normal. However, the FWU do not apply the firmware. This means that the HEX file in production will require the --pad otherwise the TF-M FWU won't work. The NS BIN file do not need HEADER and TRAILER and must not be a CONFIRMed image. CC: @etienne-lms

nandojve commented on modules/trusted-firmware-m/CMakeLists.txt:
```diff
@@ -444,6 +444,39 @@ if (CONFIG_BUILD_WITH_TFM)
+    set(image_alignment 1)
+    set(flash_write_block_size 1)
+
+    dt_chosen(chosen_flash PROPERTY "zephyr,flash")
```
I see, need to do some investigation.

JordanYates commented:
> 3- The HEX file that you flash in production should be a confirmed image.

Sure, that makes sense. That doesn't mean that the default generated files should be confirmed, since OTA images should obviously not be confirmed by default.
Standard Zephyr MCUboot builds generate `zephyr.bin`, `zephyr.signed.bin`, and `zephyr.signed.confirmed.bin`, perhaps the TF-M system should follow that convention?

JordanYates reviewed (changes requested):
Missing from any of the PR description, commit messages, or documentation (as far as I can see) is that this PR introduces new output files (`.bin`) with the same name as the existing files (`.hex`) with different meaning/file content.

For example, `zephyr/tfm_s_signed.hex` appears to be a signed + confirmed image, while `zephyr/tfm_s_signed.bin` is only a signed image. How is a user supposed to know this without digging into these cmake files?

For non TF-M applications, Zephyr explicitly separates these concepts by generating `zephyr.signed.bin` and `zephyr.signed.confirmed.bin`, for example.

The name of the files are `_signed.hex`. If you want confirmed images as well, they should be `_signed_confirmed.hex`. The file names should not lie.

JordanYates commented:
> So, in your view, the only necessary change is to rename the output file from _signed.hex to _signed_confirmed.hex. Is that correct?

It is one required change. I suspect that it would trigger other changes as a natural follow on, e.g. which files are used to generate `tfm_merged.hex`.

But if this PR is only adding new output files with different signing options, then I wouldn't have further objections.

nandojve commented:
rebase due to #93520

carlescufi commented:
Added a temporary DNM until @tomi-font or @Vge0rge have had a chance to take a look

JordanYates reviewed (dismissed):
At a minimum, the migration guide needs to note that the various hex files are now confirmed images.

Personally, I would also like to see the hex files renames to `_signed_confirmed.hex` to match the standard MCUboot pattern. The current implementation is still ending up with `image.hex` and `image.bin` having fundamentally different properties, with no indication that is the case without looking at the cmake.

nandojve commented:
Hi @tomi-font ,

> > Added a temporary DNM until @tomi-font or @Vge0rge have had a chance to take a look
> 
> @Vge0rge had a look and doesn't see issues for us but in addition I would rather trust the reviews of @JordanYates / @wearyzen on this.

Thank you for the feedback.

@wearyzen already mentioned here that he don't see any problems with the PR and now I need to address the points from @JordanYates.

<img width="920" height="577" alt="image" src="https://github.com/user-attachments/assets/2606268e-5566-495b-a511-6e3cf3cd30a3" />

So due to this and Nordic is OK, I'll remove the DNM label so I can continue with the PR.

carlescufi commented:
@nandojve 

> So due to this and Nordic is OK, I'll remove the DNM label so I can continue with the PR.

Yes, sorry I didn't remove it myself, I just got confirmation that the patch is fine from our perspective.

etienne-lms commented on modules/trusted-firmware-m/CMakeLists.txt:
```diff
@@ -444,115 +444,230 @@ if (CONFIG_BUILD_WITH_TFM)
+      message(WARNING
+        "The chosen_flash property is not defined!
+        Using flash_write_block_size and flash_erase_block_size default values
+        possible differs from TF-M board definitions resulting in improver sign."
```
```suggestion
        that may differ from TF-M board definitions resulting in invalid signatures."

etienne-lms reviewed (commented):
The change looks good to me, at least for the generic part and the ST boards.
Some minor comments.

nandojve committed:
docs: services: tfm: Add note about confirmed images

Extend the Signing Images section in the build documentation to
highlight the differences between confirmed and unconfirmed images
with respect to the PSA Certified Firmware Update API.

Signed-off-by: BUDKE Gerson Fernando <gerson.budke@leica-geosystems.com>

nandojve committed:
scripts: build: mergehex: Add --output-bin parameter

The optional --output-bin parameter instructs the IntelHex class to
save the content as a binary file instead of the Intel Hex format.

Signed-off-by: BUDKE Gerson Fernando <gerson.budke@leica-geosystems.com>

nandojve committed:
trusted-firmware-m: Prepare to generate tfm_merged.bin

When CONFIG_TFM_MCUBOOT_IMAGE_NUMBER is 1, the process to create the
final tfm_merged.bin file is more complex. This prepares the content
to introduce the generation of tfm_merged.bin for use in FOTA
applications.

Signed-off-by: BUDKE Gerson Fernando <gerson.budke@leica-geosystems.com>

nandojve committed:
trusted-firmware-m: Generate tfm_merged.bin

When CONFIG_TFM_MCUBOOT_IMAGE_NUMBER is 1, all images are merged.
Currently, there is no tfm_merged.bin file for use in FOTA. This
adds file generation to fulfill that requirement.

Signed-off-by: BUDKE Gerson Fernando <gerson.budke@leica-geosystems.com>

nordicjm reviewed (commented):
comment is unrelated anyhow, PR needs a rebase

lindblandro reviewed (commented):
LGTM :+1:

JarmouniA reviewed (commented):
There are some hard-to-ignore typos and formulation problems in commit messages and comments adeed in this PR, please request a Copilot (or any other tool) pass on those & fix the issues found, thanks!

JarmouniA commented on boards/arm/mps2/mps2_an521_cpu0_ns.dts:
```diff
@@ -107,19 +108,49 @@
+			#size-cells = <1>;
+
+			/* This code memory region must match what the TF-M
+			 * project has defined for that board - the milti image
```
```suggestion
			 * project has defined for that board - the multi-image
```

JarmouniA commented on doc/services/tfm/build.rst:
```diff
@@ -61,6 +61,12 @@ When :kconfig:option:`CONFIG_TFM_BL2` is set to ``y``, TF-M uses a secure bootlo
 
+During the signing procedure, all HEX files are marked as ``confirmed``,
+whereas all BIN files remain ``unconfirmed``. This guarantees that any image
+flashed onto a device possesses the required properties for compatibility
```
```suggestion
flashed into a device possesses the required properties for compatibility
```

JarmouniA commented on modules/trusted-firmware-m/CMakeLists.txt:
```diff
@@ -450,115 +450,254 @@ if (CONFIG_BUILD_WITH_TFM)
+      dt_prop(flash_erase_block_size PATH ${chosen_flash} PROPERTY erase-block-size)
+    else()
+      message(WARNING
+        "The chosen_flash property is not defined!
```
```suggestion
        "The 'zephyr,flash' chosen property is not defined!
```
